### PR TITLE
feat: added v3 docs

### PR DESCRIPTION
**Pull request by ChenRen93**
Allowing Tradetrust creator to issue OpenAttestation version 3 document.
https://www.pivotaltracker.com/story/show/177071031

### CODE REVIEW

File: src/common/hook/useQueue/utils/publish.tsx [hunk 1]
```diff
@@ -1,9 +1,13 @@
-import { wrapDocuments } from "@govtechsg/open-attestation";
+import {
+  wrapDocuments,
+  __unsafe__use__it__at__your__own__risks__wrapDocuments,
+  utils,
```
Comment (simboonlong) [inline src/common/hook/useQueue/utils/publish.tsx:4]:
```suggestion
  wrapDocuments,
  __unsafe__use__it__at__your__own__risks__wrapDocuments as wrapDocumentsV3,
  utils,
```

maybe can alias so can know that's v3 wrap

File: src/common/hook/useQueue/utils/publish.tsx [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
+} from "@govtechsg/open-attestation";
 import { defaultsDeep, groupBy } from "lodash";
 import { identifyProofType } from "../../../../constants/QueueState";
 import { ActionsUrlObject, Config, DocumentStorage, FormEntry, PublishingJob, RawDocument } from "../../../../types";
 import { getQueueNumber } from "../../../API/storageAPI";
-import { encodeQrCode } from "../../../utils";
+import { encodeQrCode, isV3Document } from "../../../utils";
 import { Signer } from "ethers";
```
Comment (simboonlong) [inline src/common/hook/useQueue/utils/publish.tsx:11]: see if can reuse any utils at oa level -> https://github.com/Open-Attestation/open-attestation/blob/master/src/shared/utils/guard.ts#L17

File: src/common/hook/useQueue/utils/publish.tsx [hunk 3]
```diff
@@ -61,10 +65,19 @@ export const getRawDocuments = async (forms: FormEntry[], config: Config): Promi
...
       defaultsDeep(formData, formDefaults);
-      const contractAddress =
-        formData.issuers[0]?.identityProof?.type === identifyProofType.DnsDid
-          ? identifyProofType.DnsDid
-          : formData.issuers[0]?.documentStore || formData.issuers[0]?.tokenRegistry;
+
+      let contractAddress = "";
+      if (isV3Document(formData)) {
+        contractAddress =
+          formData.openAttestationMetadata.identityProof.type.toString() == identifyProofType.DnsDid
```
Comment (simboonlong) [inline src/common/hook/useQueue/utils/publish.tsx:72]:
```suggestion
          formData.openAttestationMetadata.identityProof.type.toString() === identifyProofType.DnsDid
```

File: src/common/hook/useQueue/utils/publish.tsx [hunk 4]
```diff
@@ -100,14 +113,20 @@ export const groupDocumentsIntoJobs = async (
...
-    const wrappedDocuments = wrapDocuments(rawDocuments);
+
+    const wrappedDocuments = isV3Document(rawDocuments[0])
+      ? await __unsafe__use__it__at__your__own__risks__wrapDocuments(rawDocuments)
+      : wrapDocuments(rawDocuments);
+
     const firstWrappedDocument = wrappedDocuments[0];
+    const merkleRoot = utils.isWrappedV3Document(firstWrappedDocument)
+      ? firstWrappedDocument.proof.merkleRoot
+      : firstWrappedDocument.signature?.merkleRoot;
```
Comment (simboonlong) [inline src/common/hook/useQueue/utils/publish.tsx:129]: think here have already, can use this instead -> https://github.com/Open-Attestation/open-attestation/blob/master/src/shared/utils/utils.ts#L82

File: src/common/hook/useQueue/utils/publish.tsx [hunk 5]
```diff
@@ -168,5 +187,6 @@ export const getPublishingJobs = async (
 ): Promise<PublishingJob[]> => {
   // Currently works for only multiple verifiable document issuance:
   const rawDocuments = await getRawDocuments(forms, config);
-  return groupDocumentsIntoJobs(rawDocuments, nonce, signer);
+  const test = await groupDocumentsIntoJobs(rawDocuments, nonce, signer);
+  return test;
```
Comment (simboonlong) [inline src/common/hook/useQueue/utils/publish.tsx:191]: use a better variable name?

File: src/common/utils.tsx [hunk 6]
```diff
@@ -84,3 +85,8 @@ export const downloadJsonDataFile = (jsonTemplate: any): void => {
 export const isWalletOption = (option: string | WalletOptions): option is string => {
   return typeof option === "string";
 };
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export const isV3Document = (document: any): document is v3.OpenAttestationDocument => {
+  return document.hasOwnProperty("@context");
+};
```
Comment (simboonlong) [inline src/common/utils.tsx:92]: look in oa utils

File: src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx [hunk 7]
```diff
@@ -5,19 +5,21 @@ import {
...
-import { v2 } from "@govtechsg/open-attestation";
+import { v2, v3 } from "@govtechsg/open-attestation";
+import { isV3Document } from "../../../common/utils";
 
 type Dispatch = (action: HostActions) => void;
 
 interface DocumentPreview {
-  document: v2.OpenAttestationDocument;
+  document: v2.OpenAttestationDocument | v3.OpenAttestationDocument;
 }
```
Comment (simboonlong) [inline src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx:15]: reuse -> https://github.com/Open-Attestation/open-attestation/blob/master/src/shared/%40types/document.ts#L15

File: src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx [hunk 8]
```diff
@@ -5,19 +5,21 @@ import {
...
 
 export const DocumentPreview: FunctionComponent<DocumentPreview> = ({ document }) => {
   const toFrame = useRef<Dispatch>();
   const [height, setHeight] = useState(0);
-
-  const rendererTemplate = document["$template"] as v2.TemplateObject;
+  const rendererTemplate = isV3Document(document)
+    ? document["openAttestationMetadata"]["template"]
+    : (document["$template"] as v2.TemplateObject);
```
Comment (simboonlong) [inline src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx:22]: maybe can do a `getTemplate` utility at oa side, but anyway can it be dot notation format? 
```suggestion
  const rendererTemplate = isV3Document(document)
    ? document.openAttestationMetadata.template
    : (document.$template as v2.TemplateObject);
```

File: src/common/hook/useQueue/utils/publish.tsx (hunk 5 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:191]: done.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 4 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:129]: done.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 2 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:11]: the utils only support checking of wrapped and signed documents. it doesn't check if the docs is unwrapped documents, shall I add one in oa utils ?

File: src/common/hook/useQueue/utils/publish.tsx (hunk 1 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:4]: done.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 3 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:72]: done.

File: src/common/utils.tsx (hunk 6 above)
Comment (ChenRen93) [reply src/common/utils.tsx:92]: The utils only support checking of wrapped and signed documents.

File: src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx (hunk 7 above)
Comment (ChenRen93) [reply src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx:15]: oh, done.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 2 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:11]: Updated, change to isRawV3Document in OA utils

File: src/common/utils.tsx (hunk 6 above)
Comment (ChenRen93) [reply src/common/utils.tsx:92]: removed, all usage of this function will change to isRawV3Document in OA utils

File: src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx (hunk 8 above)
Comment (ChenRen93) [reply src/components/DynamicFormContainer/DocumentPreview/DocumentPreview.tsx:22]: removed, change to retrieve getTemplateURL in oa utils

File: src/common/hook/useQueue/utils/publish.tsx (hunk 3 above)
Comment (simboonlong) [reply src/common/hook/useQueue/utils/publish.tsx:72]: bump*

File: src/test/fixtures/sample-config-ropsten-v3.json [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    {
+      "name": "Testing Version 3",
+      "type": "VERIFIABLE_DOCUMENT",
+      "defaults": {
+        "@context": [
+          "https://www.w3.org/2018/credentials/v1",
+          "https://schemata.openattestation.com/com/openattestation/1.0/DrivingLicenceCredential.json",
+          "https://schemata.openattestation.com/com/openattestation/1.0/OpenAttestation.v3.json",
+          "https://schemata.openattestation.com/com/openattestation/1.0/CustomContext.json"
+        ],
```
Comment (simboonlong) [inline src/test/fixtures/sample-config-ropsten-v3.json:17]: no need `CustomContext` i think, can put in invoice and coo forms too 🎉

File: integration/document-data-upload.spec.ts [hunk 10]
```diff
@@ -2,26 +2,36 @@ import { Selector } from "testcafe";
 import { enterPassword, loadConfigFile } from "./helper";
 import { join } from "path";
 import { homedir } from "os";
-import { existsSync, read, readFileSync } from "fs";
+import { existsSync, read, readFileSync, unlinkSync, unwatchFile, watchFile } from "fs";
```
Comment (cavacado) [inline integration/document-data-upload.spec.ts:5]: remove unused imports

File: src/common/hook/useQueue/utils/publish.tsx [hunk 11]
```diff
@@ -61,10 +65,19 @@ export const getRawDocuments = async (forms: FormEntry[], config: Config): Promi
...
+        contractAddress =
+          formData.openAttestationMetadata.identityProof.type.toString() === identifyProofType.DnsDid
+            ? identifyProofType.DnsDid
+            : formData.openAttestationMetadata.proof.value;
+      } else {
+        contractAddress =
+          formData.issuers[0]?.identityProof?.type === identifyProofType.DnsDid
+            ? identifyProofType.DnsDid
+            : formData.issuers[0]?.documentStore || formData.issuers[0]?.tokenRegistry;
+      }
```
Comment (cavacado) [inline src/common/hook/useQueue/utils/publish.tsx:80]: make these lines into a local helper function `getContractAddressFromRawDoc`

File: src/common/hook/useQueue/utils/publish.tsx [hunk 12]
```diff
@@ -100,14 +113,19 @@ export const groupDocumentsIntoJobs = async (
   let nonce = currentNonce;
 
   const jobs: PublishingJob[] = [];
-
   // Process all verifiable documents with document store first
-  documentStoreAddresses.forEach((contractAddress) => {
+  for (const contractAddress of documentStoreAddresses) {
```
Comment (cavacado) [inline src/common/hook/useQueue/utils/publish.tsx:117]: was wondering why switch to `for .. of` from `fromEach`

File: src/common/hook/useQueue/utils/publish.tsx [hunk 13]
```diff
@@ -143,19 +163,24 @@ export const groupDocumentsIntoJobs = async (
...
 
   // Process all transferable records next
-  transferableRecords.forEach((transferableRecord) => {
+  for (const transferableRecord of transferableRecords) {
     const { type, contractAddress, rawDocument, payload } = transferableRecord;
-    const documents = wrapDocuments([rawDocument]);
+
+    const documents = utils.isRawV3Document(rawDocument)
+      ? await wrapDocumentsV3([rawDocument])
+      : wrapDocuments([rawDocument]);
```
Comment (cavacado) [inline src/common/hook/useQueue/utils/publish.tsx:171]: you've used this pattern thrice already, might be good to make it into a simple helper function `wrapDocuments`, can alias the previous `wrapDocuments` to `wrapDocumentsV2`.

File: src/common/hook/useQueue/utils/publish.tsx [hunk 14]
```diff
@@ -168,5 +193,5 @@ export const getPublishingJobs = async (
 ): Promise<PublishingJob[]> => {
   // Currently works for only multiple verifiable document issuance:
   const rawDocuments = await getRawDocuments(forms, config);
-  return groupDocumentsIntoJobs(rawDocuments, nonce, signer);
+  return await groupDocumentsIntoJobs(rawDocuments, nonce, signer);
```
Comment (cavacado) [inline src/common/hook/useQueue/utils/publish.tsx:196]: is this necessary?

**Review by cavacado — COMMENTED** [review]
i'll need to try to create a doc first but for now here are my preliminary comments

File: src/common/hook/useQueue/utils/publish.tsx (hunk 12 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx:117]: cause wrapDocumentsV3 is async, and require the wrapped doc to be pushed into the jobs array before returning. thus it is easier to implement async/await using for..of compare to forEach

Comment (cavacado) [reply src/common/hook/useQueue/utils/publish.tsx:117]: hmm was `wrapDocumentsV2` sync previously? i was under the impression it was always async (ah its not, just checked).

File: integration/document-data-upload.spec.ts (hunk 10 above)
Comment (ChenRen93) [reply integration/document-data-upload.spec.ts]: removed, thanks.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 11 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx]: okay extracted.

File: src/common/hook/useQueue/utils/publish.tsx (hunk 13 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx]: okay extracted into local helper function

File: src/common/hook/useQueue/utils/publish.tsx (hunk 14 above)
Comment (ChenRen93) [reply src/common/hook/useQueue/utils/publish.tsx]: nope, i have removed. thanks

**Review by cavacado — APPROVED** [review]
looks good, if e2e tests pass im okay 🎊

**Pull request merged**